196 data providers

raphaeltraviss opened the pull request:
Added data provider for variables, and switched dbQuery to use drupal_write_record().

raphaeltraviss committed:
Added data provider for variables.

raphaeltraviss committed:
Activate example render cache for tests.

raphaeltraviss committed:
Added example watchdog resource and drupal_write_record.

raphaeltraviss committed:
Fixed dbQuery test.

raphaeltraviss committed:
Removed debug statements.

raphaeltraviss committed:
Moved example plugins to correct location.

e0ipso commented on modules/restful_example/plugins/restful/db_query/watchdog/1.0/RestfulWatchdogResource.class.php:
```diff
@@ -0,0 +1,48 @@
+ */
+
+/**
+ * An abstract implementation of RestfulEntityInterface.
```
This class is not abstract ;-)

e0ipso commented on modules/restful_example/plugins/restful/db_query/watchdog/1.0/watchdog.inc:
```diff
@@ -0,0 +1,15 @@
+  'description' => t('Expose watchdog entries to the REST API.'),
+  'class' => 'RestfulWatchdogResource',
+  'authentication_types' => TRUE,
+  'authentication_optional' => TRUE,
```
Let's remove the `authentication_optional`.

e0ipso replied:
~~Actually we should add an access check to this endpoint, since it can contain sensitive information. We wouldn't want a security issue because someone left `restful_example` enabled.~~

I see that you already did that :-)

e0ipso commented on plugins/restful/RestfulDataProviderDbQuery.php:
```diff
@@ -99,7 +113,7 @@ public function __construct(array $plugin, \RestfulAuthenticationManager $auth_m
     $this->tableName = $options['table_name'];
     $this->idColumn = $options['id_column'];
-    $this->primary = empty($plugin['primary']) ? NULL : $this->primary = $plugin['primary'];
+    $this->primary = !empty($plugin['primary']) ? $options['primary'] : $options['id_column'];
```
What's this for?

raphaeltraviss replied:
This will set `$primary` from the id column, in case the `'primary'` key is omitted from the plugin definition.  This is used in `update()` to skip updating the primary and id fields.

e0ipso commented on plugins/restful/RestfulDataProviderDbQuery.php:
```diff
@@ -341,37 +355,35 @@ protected function replace($id) {
       }
-      // Check if the public property is set in the payload.
-      if (!isset($request[$public_property])) {
-        if ($full_replace) {
```
Are we losing this functionality?

`$full_replace` is what separates a `PUT` request from a `PATCH` request.

raphaeltraviss replied:
My thinking here was that `drupal_write_record()` always performs a partial update.  But, come to think of it, if we are doing `$full_replace`, we may be able to feed it empty values for the non-supplied fields to get it to work for a `PUT` context.  I'll look into this.

raphaeltraviss replied:
Yes, this works, adding in next commit.

e0ipso commented on plugins/restful/RestfulDataProviderDbQuery.php:
```diff
@@ -341,37 +355,35 @@ protected function replace($id) {
-    // Once the update array is built, execute the query.
-    $query->fields($fields)->execute();
+    // Add the id column values into the record.
+    foreach ($this->getIdColumn() as $index => $column) {
```
~~Throw an exception if `$this->getIdColumn()` is empty.~~

Wait, that won't happen.

mateu-aguilo-bosch replied:
Since it would become an insert in an update function call.

e0ipso commented on plugins/restful/RestfulDataProviderDbQuery.php:
```diff
@@ -341,37 +355,35 @@ protected function replace($id) {
+
+    // Once the record is built, write it.
+    if (!drupal_write_record($this->getTableName(), $record, $id_columns)) {
+      throw new \RestfulServiceUnavailable('Record could not be updated to the database.');
```
:+1:

e0ipso commented on plugins/restful/RestfulDataProviderDbQuery.php:
```diff
@@ -379,51 +391,38 @@ public function update($id, $full_replace = FALSE) {
     ));
-    return $this->view($id, TRUE);
+
+    // @todo: do we need to re-form $id in case some of the values have changed?
```
I don't think we need. What scenario are you thinking about?

e0ipso replied:
Either way, we need to create a follow up issue in GitHub and remove the `@todo` in code or just remove the comment.

raphaeltraviss replied:
This was a left over comment from before I understood how the id processing worked; will be deleted.

e0ipso replied:
Maybe you could drop a comment on how it works with your own words, it will help the next time we revisit this.

e0ipso commented on plugins/restful/RestfulDataProviderDbQuery.php:
```diff
@@ -379,51 +391,38 @@ public function update($id, $full_replace = FALSE) {
 
-    // Once the update array is built, execute the query.
-    if ($id = $query->fields($fields)->execute()) {
       return $this->view($id, TRUE);
```
Do we want to just: `$this->view($id);`?

e0ipso commented on plugins/restful/RestfulDataProviderVariable.php:
```diff
@@ -0,0 +1,306 @@
+   * @param string $language
+   *   (optional) The language to return items in.
+   */
+  public function __construct(array $plugin, \RestfulAuthenticationManager $auth_manager = NULL, \DrupalCacheInterface $cache_controller = NULL, $language = NULL) {
```
If we're just calling the parent, then we don't need to define the constructor.

e0ipso commented on plugins/restful/RestfulDataProviderVariable.php:
```diff
@@ -0,0 +1,306 @@
+  /**
+   * {@inheritdoc}
+   */
+  public static function controllersInfo() {
```
Is there any need to re-define the controller names? Can't we just use the usual ones? I think it helps with setting expectation on what's in a data provider class.

e0ipso commented on plugins/restful/RestfulDataProviderVariable.php:
```diff
@@ -0,0 +1,306 @@
+      if (isset($public_fields[$public_field_name]['property'])) {
+        $property_name = $public_fields[$public_field_name]['property'];
+        if ($property_name == 'name') {
+          uksort($variables, array(__CLASS__, 'nameSort'));
```
I didn't know that you could use the keyword `__CLASS__` as a substitute for `$this` in the callable context :-)

e0ipso commented on plugins/restful/RestfulDataProviderVariable.php:
```diff
@@ -0,0 +1,306 @@
+   * @param string $b
+   *   The second variable name for the comparison.
+   */
+  public function nameSort($a, $b) {
```
If this is not using the class' state, can we make it an static method?

e0ipso commented on plugins/restful/RestfulDataProviderVariable.php:
```diff
@@ -0,0 +1,306 @@
+   *   The second data structure for the comparison.
+   */
+  public function valueSort($a, $b) {
+    return;
```
Are we not sorting?

raphaeltraviss replied:
By default, we don't do any sorting on the variable values, since their data structures can be completely different.  This method exists as a placeholder, to be overridden by the plugin author.  If they have some specific variables that they know the data structure of, and they want to sort by, then they can override this method in their plugin.  Maybe I should replace the algorithm comment with something more useful?

e0ipso replied:
:+1: to adding a `@code` implementation example in the docblock.

e0ipso commented on plugins/restful/RestfulDataProviderVariable.php:
```diff
@@ -0,0 +1,306 @@
+        if ($property_name == 'name') {
+          uksort($variables, array(__CLASS__, 'nameSort'));
+        }
+        elseif ($property_name == 'value') {
```
Instead of an `if/else`, would it be possible to have a generic handle of this that would not need to separate by `$property_name`?

Maybe having a generic `$this->stringCompare` method?

raphaeltraviss replied:
Yes, it could go either way.  But I think the tie-breaker is that fact that you need to know the properties in order to know if you should run the sort function on the keys, or the values of the array, i.e. `uksort()` vs `uasort()`.  And, since we need to know which property we're sorting, that allows us to break them out into two separate methods: `nameSort()`, which is essentially string comparison, and `valueSort()`, which compares data structures in some specific way.

e0ipso commented on plugins/restful/RestfulDataProviderVariable.php:
```diff
@@ -0,0 +1,306 @@
+   * {@inheritdoc}
+   */
+  public function getTotalCount() {
+    return array('restful_count' => count($this->getVariablesForList()));
```
The count should not include pagination. This should be something like `count($GLOBALS['conf'])`.

Also, I'm not sure I get why you are returning an array. This is what I see for other data providers:

``` php
  /**
   * Helper method to get the total count of entities that match certain
   * request.
   *
   * @return int
   *   The total number of results without including pagination.
   */
  public function getTotalCount();
```

e0ipso commented on plugins/restful/RestfulDataProviderVariable.php:
```diff
@@ -0,0 +1,306 @@
+ * Contains \RestfulDataProviderDbQuery
+ */
+
+abstract class RestfulDataProviderVariable extends \RestfulBase implements \RestfulDataProviderInterface {
```
Maybe we need to create a `RestfulDataProviderVariableInterface`. Just like we have `RestfulDataProviderCToolsPluginsInterface`.

e0ipso commented on plugins/restful/RestfulDataProviderVariable.php:
```diff
@@ -0,0 +1,306 @@
+    if (empty($names)) {
+      return array();
+    }
+    $variables = array_keys($this->getVariablesForList());
```
`getVariablesForList` will return the paginated list, we should access the complete list of variables. `global $conf` instead?

raphaeltraviss replied:
Question: is `viewMultiple()` expected to obey the sort order of the passed-in `$names`, or the sort defined by the plugin?

e0ipso replied:
There's no guarantee that the output order is the same of the order of the ids. This means that the output order is the one defined by the plugin.

e0ipso commented on plugins/restful/RestfulDataProviderVariable.php:
```diff
@@ -0,0 +1,306 @@
+    $output = array();
+
+    // Build output according to filtered/sorted list of variables.
+    foreach ($variables as $variable) {
```
Can we revers the order of `foreach`? There are potentially lots and lots of variables. Possibly something like:

``` php
global $conf;
foreach ($names as $name) {
  if (!empty($conf[$name])) {
    $output[] = $this->view($variable);
  }
}
```

raphaeltraviss replied:
That is my attempt at sorting logic; I looped through `$variables` because they were in the correct sort order from `getVariablesForList()`.  So, this depends on whether we want this method to sort according to the plugin definition, or the passed-in `$names`.

e0ipso replied:
Can we build the array and then pass the results through `uasort`?

e0ipso commented on plugins/restful/RestfulDataProviderVariable.php:
```diff
@@ -0,0 +1,306 @@
+        $public_field_value = variable_get($name);
+      }
+      else {
+        continue;
```
Throw an exception explaining that the only possible properties for the variable resource are: `'name'` and `'value'`. Only if the `$info['property']` is populated.

e0ipso commented on plugins/restful/RestfulDataProviderVariable.php:
```diff
@@ -0,0 +1,306 @@
+      }
+
+      // Modify the public field value using a callback, if supplied.
+      $public_field_value = $info['callback'] ? static::executeCallback($info['callback'], array($value)) : $public_field_value;
```
The `$info['callback']` provides an way to wire a callback to your public field. This should be something like:

``` php
      if ($info['callback']) {
        $public_field_value = static::executeCallback($info['callback'], array($name));
      }
```

e0ipso commented on plugins/restful/RestfulDataProviderVariable.php:
```diff
@@ -0,0 +1,306 @@
+   *  The name of the variable to set a value for.
+   */
+  public function update($name, $full_update = TRUE) {
+    $this->setVariable($name);
```
We need to use the `$full_update` in order to tell `PUT` and `PATCH` methods apart.

With the `PUT` method if we don't provide a value, then NULL is assumed to be the value to set. With `PATCH` if we don't provide a value then the current value is not touched.

e0ipso commented on plugins/restful/RestfulDataProviderVariable.php:
```diff
@@ -0,0 +1,306 @@
+      }
+    }
+    if (isset($name)) {
+      variable_set($name, $value);
```
It's possible that there is no `$value` set (see above for `PUT` vs `PATCH`)

e0ipso commented on plugins/restful/RestfulDataProviderVariable.php:
```diff
@@ -0,0 +1,306 @@
+      return $this->view($name);
+    }
+    else {
+      throw new Exception('No name property supplied');
```
The RESTful module always throws its own exceptions so we can do fancy things with them later. In this case you should throw a `RestfulBadRequestException`. Although in this case we don't need it since we will always have a name (because it's the ID for the update method).

raphaeltraviss replied:
My thinking was that we needed the exception for `POST` requests, where `name` might not have been sent in the body, since we'd have nothing to create and nothing to view.

EDIT: ah, ok, `create` contexts set the name from the payload, `update` and `replace` provide it as part of the method call.

e0ipso commented on plugins/restful/RestfulDataProviderVariable.php:
```diff
@@ -0,0 +1,306 @@
+   *
+   * @todo: get $name from the request: make it optional for create()/update().
+   */
+  public function setVariable() {
```
`public function setVariable($name) {`

e0ipso commented on plugins/restful/RestfulDataProviderVariable.php:
```diff
@@ -0,0 +1,306 @@
+   * @param string $name
+   *  The name of the variable to set a value for.
+   *
+   * @todo: get $name from the request: make it optional for create()/update().
```
`$name` is ignored in the payload for the update (because it's the ID of the resource). Although it should be mandatory on create.

Please remove the `@todo` comment.

raphaeltraviss replied:
Another left-over, will remove.

mateu-aguilo-bosch commented on tests/RestfulDbQueryTestCase.test:
```diff
@@ -68,19 +65,20 @@ class RestfulDbQueryTestCase extends DrupalWebTestCase {
     $result = $handler->get($id);
-    // Remove nullified fields.
-    $result = array_filter($result[0]);
-    $this->assertEqual($result, $mock_data2, 'The record was updated with PUT successfully.');
```
Why do we not need this anymore?

e0ipso commented on tests/modules/restful_test/restful_test.install:
```diff
@@ -54,5 +54,37 @@ function restful_test_schema() {
     'primary key' => array('pid'),
   );
 
+  // Defining table via hook_install() due to drupal_write_record().
```
Does this mean that if we are not declaring the table via `hook_schema` we cannot use `drupal_write_record`?

raphaeltraviss replied:
Correct.  `drupal_write_record()` uses `drupal_get_schema()` to find the names of the fields for the supplied table, and if the table is not defined in the schema, it will return `FALSE`.

amitaibu replied:
:+1:

e0ipso commented on tests/RestfulDbQueryTestCase.test:
```diff
@@ -68,19 +65,20 @@ class RestfulDbQueryTestCase extends DrupalWebTestCase {
-    $result = array_filter($result[0]);
-    $this->assertEqual($result, $mock_data2, 'The record was updated with PUT successfully.');
+    // Make mock data resemble the intended response.
+    $mock_data2['serialized'] = $mock_data['serialized_field'];
```
Actually this is the same `PUT` and `PATCH` situation as before. If we do not provide a _serialized_ and we call `PUT` that means that the _serialized_ should be nullified.

e0ipso replied:
Please, see that tests reflect that.

e0ipso commented on tests/RestfulVariableTestCase.test:
```diff
@@ -0,0 +1,127 @@
+   *
+   * @var string
+   */
+  protected $tableName = 'restful_test_variable';
```
Copy & paste leftover?

e0ipso replied:
I can remove this.

e0ipso commented on tests/RestfulVariableTestCase.test:
```diff
@@ -0,0 +1,127 @@
+   * Test authenticating a user.
+   */
+  function testCrudOperations() {
+    // Populate the table with some variables.
```
`// Populate with some variables.`

e0ipso commented on tests/RestfulVariableTestCase.test:
```diff
@@ -0,0 +1,127 @@
+    $results = $handler->get('variable_5');
+    $result = $results[0];
+    $this->assertEqual($result['variable_name'], 'variable_5', 'The variable name was successfully retrieved.');
+    $this->assertNotNull($result['variable_value'], 'The variable value was successfully retrieved.');
```
We should store the random number when it gets generated and compared here:

``` php
$this->assertEqual($result['variable_value'], $numbers[5], 'The variable value was successfully retrieved.');
```

e0ipso commented on tests/RestfulVariableTestCase.test:
```diff
@@ -0,0 +1,127 @@
+    $this->assertTrue($inResults, 'All the content listed successfully.');
+
+    // Testing sort.
+    variable_set('zzz', array('some_property' => $this->randomName()));
```
Imagine a site that has a variable named `zzz_counter` (for the imaginary zzz module). This can lead to false positives. We should get all the variable names, then get the last (in strcmp order) and generate one that would compare after that.

e0ipso commented on tests/RestfulVariableTestCase.test:
```diff
@@ -0,0 +1,127 @@
+    $results = $handler->get('created_variable');
+    $initial_value = $results[0]['variable_value'];
+    $this->assertEqual($results[0]['variable_name'], 'created_variable', 'The variable was created.');
+    $this->assertTrue($initial_value, 'The created variable value is present.');
```
``` php
$this->assertEqual($initial_value, $request['variable_value'], 'The variable was successfully created.');
```

e0ipso commented:
@rfx1007 this is wonderful work!

Thanks a lot for this contribution, I can only imagine all the effort you put for this pull request. I made a thorough peer review out of respect for your awesome work.

You'll see that there are some questions, some suggestions and some clarifications. If you don't feel like implementing those, let me know and I'll pop in a PR on top of your work.

![respect](http://i.giphy.com/CcNL1QhM1b3ck.gif)

raphaeltraviss commented:
Thanks for the in-depth review; I'll take care of the remaining issues and complete the pull request.

e0ipso commented:
@rfx1007 You are welcome!

raphaeltraviss committed:
Code improvement as per first review of pull request 356.

raphaeltraviss committed:
Merge branch '7.x-1.x' of https://github.com/Gizra/restful into 196-data_providers

Conflicts:
	plugins/restful/RestfulDataProviderDbQuery.php
	tests/RestfulDbQueryTestCase.test
	tests/modules/restful_test/restful_test.module

raphaeltraviss committed:
Removed git conflict.

raphaeltraviss committed:
Removed yet more git conflicts.

raphaeltraviss committed:
Changed test group Restful to RESTful.

raphaeltraviss committed:
Trying to get tests working.

raphaeltraviss committed:
Fixed DbQuery count result.

rfx1007 commented:
I've implemented all of the items from the code review, but now, many of the tests are failing: I've got about 75 exceptions and 15 failures in Node Create, CSRF, and others.  Many of the exceptions are from `entity_metadata_no_hook_node_access()`.  I've tried using entity API dev version with the two patches applied, and entity API from git with the one patch it still needs, same result.  Any hints as to what might be wrong?

raphaeltraviss committed:
Fix my botched merge 2aa31ef for passthrough handling.

raphaeltraviss committed:
Merge branch '7.x-1.x' of https://github.com/Gizra/restful into 196-data_providers

raphaeltraviss committed:
Renamed function to try and get Travis to complete the test.

raphaeltraviss committed:
Removed field creation function from setup function.

raphaeltraviss commented:
Test lights are finally green!
- removed variable value sort, since it was a weird concept to begin with, and could not possibly be supplied via URL
- the variable array is now indexed, with each element containing the property, similar to how the other data providers work.  Reworked view/viewMultiple
- integrated passthrough commit with dbquery provider and got the tests working

e0ipso commented on modules/restful_token_auth/tests/RestfulTokenAuthenticationTestCase.test:
```diff
@@ -11,13 +11,12 @@ class RestfulTokenAuthenticationTestCase extends DrupalWebTestCase {
       'name' => 'Token Authentication',
       'description' => 'Test the request authentication with a token.',
-      'group' => 'Restful',
+      'group' => 'RESTful',
```
:+1:

e0ipso commented:
Merged!

Many thanks for this great contribution. I'll make sure to mention your effort in the upcoming release notes.

amitaibu commented:
Thanks @rfx1007 !

raphaeltraviss commented:
Whew!  Finally bagged this one!  Time for a drink.
